Ensure webhook is started

kajinamit opened the pull request:
This ensures webhook is started when controller-manage deployment becomes ready.

gibizer commented:
The envtest fails as it does not find the keystone CRD any more via the api package. We might need to re-think how we get those CRDs in the test env now that the full package is not in our dependency but the CRD is only available in the full package. I will try to get back to this today to do a deeper look...

stuggi commented:
> The envtest fails as it does not find the keystone CRD any more via the api package. We might need to re-think how we get those CRDs in the test env now that the full package is not in our dependency but the CRD is only available in the full package. I will try to get back to this today to do a deeper look...

yes, I have also started looking/thinking about it and testing right now:
- keep the list as we had it
- get the version of the api submodule we have in go.mod
- in envsetup install the module with version from above

gibizer commented:
I think the main question we have to resolve is: How can we get the correct version of the CRD based on the version of the keystone/api module. I believe that now the keystone and the keystone/api modules are (can be) versioned separately. And now placement states correctly only the version of keystone/api it depends on. So from either:

-a) Deduce the version of the keystone module from the version of the keystone/api version in our go.mod and install that version of the keystone module in the test env to get the CRD. I'm not sure this is possible as there might not be 1:1 relationship between keystone/api and keystone module versions

- b) move the keystone CRD somehow to the keystone/api module so that we can get it from there. This would keep the API struct definitions and the generated CRD at the same module which is nice. But I'm not sure this can be done with operator-sdk.

-c) we take the API structs from the keystone/api module and generate the CRD locally. I have to look into how to trigger such regeneration.

SeanMooney commented:
do me the CRDs are the public user facing API of the operator so it makes sense to me that they would be version with the go api and be part of the same module.

so if we can do b i would

gibizer commented:
My only idea for #b) is to simlink the /config dir into /api/ . @stuggi, @dprince: could this be a way forward?

For #c) I  can generate the CRD locally from the API structs with:
``` bin/controller-gen crd paths=/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/keystone-operator/api@v0.0.0-20220913134025-daee844b58f7/v1beta1 output:crd:dir=/tmp/crds output:stdout``` We already generate the path to the locally downloaded keystone/api module. So we only need an extra step to run the generator. I will try to hack something up for it.

SeanMooney commented:
i guess since the  crds are generated form the go code C is mostly the same a b so ya we could do that in the envtest case.

stuggi commented:
I was testing a) but that is really hackish and if you want to use replace not easy to use. So I stopped it.

> My only idea for #b) is to simlink the /config dir into /api/ . @stuggi, @dprince: could this be a way forward?
could just symlink config/crd/base and could add a check in the `make manifests` that it is there.

> 
> For #c) I can generate the CRD locally from the API structs with: ` bin/controller-gen crd paths=/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/keystone-operator/api@v0.0.0-20220913134025-daee844b58f7/v1beta1 output:crd:dir=/tmp/crds output:stdout` We already generate the path to the locally downloaded keystone/api module. So we only need an extra step to run the generator. I will try to hack something up for it.

I think b) could work (have not tested) and might be the quickest. c) has the same end result, so could also do that.

stuggi commented:
this is to demo the symlink approach. basically each operator need an update to move `/config/crd/bases` to `/api/bases` and create a link `/config/crd/bases` -> `/api/bases`. I also added a small enhancement so that replace is also checked in go.mod for a  module. Like in a commit of this PR I point to my private clone for keystone/mariadb where I did the above link change.

gibizer commented on controllers/suite_test.go:
```diff
@@ -67,31 +67,46 @@ func TestAPIs(t *testing.T) {
+
+	// check for replacement config in go.mod for the named module
+	for _, r := range f.Replace {
+		if r.Old.Path == moduleName {
```
I'm mistaken. This now finds all the replacement and take the last one. Regardless it seems to be OK as there can only one such replacement for a given module

gibizer replied:
I think this is OK as is. I also believe that we cannot have two replacements for the same module

gibizer commented on controllers/suite_test.go:
```diff
@@ -67,31 +67,46 @@ func TestAPIs(t *testing.T) {
+		return name, version, nil
+	}
+
+	return name, "", fmt.Errorf("Cannot find %s in our go.mod file", moduleName)
```
Probably does not matter. I'm just curious what is the golang recommendation for returning an error with a partial result?

gibizer replied:
Thanks for the explanation.

gibizer commented:
> this is to demo the symlink approach. basically each operator need an update to move `/config/crd/bases` to `/api/bases` and create a link `/config/crd/bases` -> `/api/bases`. I also added a small enhancement so that replace is also checked in go.mod for a module. Like in a commit of this PR I point to my private clone for keystone/mariadb where I did the above link change.

If we can agree on this pattern for our operators then simlinking seems to be the clean approach. 
(Generating locally has the issue that I would need to run the generator from the test code to be able to ensure that the generator sees the correct dependency versions)

/lgtm

stuggi commented:
> > this is to demo the symlink approach. basically each operator need an update to move `/config/crd/bases` to `/api/bases` and create a link `/config/crd/bases` -> `/api/bases`. I also added a small enhancement so that replace is also checked in go.mod for a module. Like in a commit of this PR I point to my private clone for keystone/mariadb where I did the above link change.
> 
> If we can agree on this pattern for our operators then simlinking seems to be the clean approach. (Generating locally has the issue that I would need to run the generator from the test code to be able to ensure that the generator sees the correct dependency versions)
> 
> /lgtm

@dprince what is your take on the symlink change in the operators?

dprince commented:
My vote: store all the required CR's locally like we do for routev1: https://github.com/openstack-k8s-operators/placement-operator/blob/master/controllers/suite_test.go#L112

stuggi commented:
> Is there a way to have the test pull the version of the keystone CRD from the go.mod version?

I had a quick look at that, but was not fully working. I'll revisit that again that we can compare.

gibizer commented:
> Is there a way to have the test pull the version of the keystone CRD from the go.mod version?

The go.mod of placement only has the keystone-operator/api  not keystone-operator any more, so we can only pull keystone/api,  but that module does not have the CRD in it (hence the symlinking).

stuggi commented:
> > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> 
> The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).

I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.

stuggi commented:
> > > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> > 
> > 
> > The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).
> 
> I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.

@gibizer @dprince @SeanMooney please have a look at https://github.com/openstack-k8s-operators/placement-operator/pull/30

gibizer commented:
> > > > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> > > 
> > > 
> > > The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).
> > 
> > 
> > I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.
> 
> @gibizer @dprince @SeanMooney please have a look at #30

I still think that storing (or symlinking) the CRD in the same module where the API Structs are stored is the cleanest approach. Those two things belong to the same API definition and there is a very high dependency between them.

stuggi commented:
> > > > > Is there a way to have the test pull the version of the keystone CRD from the go.mod version?
> > > > 
> > > > 
> > > > The go.mod of placement only has the keystone-operator/api not keystone-operator any more, so we can only pull keystone/api, but that module does not have the CRD in it (hence the symlinking).
> > > 
> > > 
> > > I have a working patch to load the base module using the api version details of the go.mod . will submit in a bit so that we can discuss that approach.
> > 
> > 
> > @gibizer @dprince @SeanMooney please have a look at #30
> 
> I still think that storing (or symlinking) the CRD in the same module where the API Structs are stored is the cleanest approach. Those two things belong to the same API definition and there is a very high dependency between them.

yes I agree, storing in /api and add a symlink from the original place to be able to use the default create mechanism is the easiest and cleanest way. the other options PR #30 or store a copy of the required CRDs in the operator where it is needed, can lead to wrong versions if there is no corresponding base version to the sub module, or it got missed to sync them.

stuggi commented:
I'll get a new one filed now that we have https://github.com/openstack-k8s-operators/keystone-operator/pull/102 and https://github.com/openstack-k8s-operators/mariadb-operator/pull/35

kajinamit committed:
Ensure webhook is started

This ensures webhook is started when controller-manage deployment
becomes ready.

gibizer commented on tests/functional/suite_test.go:
```diff
@@ -185,15 +183,9 @@ var _ = BeforeSuite(func() {
-		}
-		conn.Close()
-		return nil
+		checker := k8sManager.GetWebhookServer().StartedChecker()
```
I'm not sure this actually does what we expect. I tried to comment out `err = (&placementv1.PlacementAPI{}).SetupWebhookWithManager(k8sManager)` above to see if this new check catches the missing webhook but it does not.

kajinamit replied:
Webhooks should be installed to use this check. This check internally executes the same connection check we currently have. See https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/webhook/server.go#L308-L327 .

gibizer replied:
Something is still different between the two solution. 

1. If on main I disable the webhook:
```
diff --git a/tests/functional/suite_test.go b/tests/functional/suite_test.go
index 750b272..05e5e3a 100644
--- a/tests/functional/suite_test.go
+++ b/tests/functional/suite_test.go
@@ -165,8 +165,8 @@ var _ = BeforeSuite(func() {
        kclient, err := kubernetes.NewForConfig(cfg)
        Expect(err).ToNot(HaveOccurred(), "failed to create kclient")
 
-       err = (&placementv1.PlacementAPI{}).SetupWebhookWithManager(k8sManager)
-       Expect(err).NotTo(HaveOccurred())
+       //      err = (&placementv1.PlacementAPI{}).SetupWebhookWithManager(k8sManager)
+       //      Expect(err).NotTo(HaveOccurred())
 
        placementv1.SetupDefaults()
```
and then run a test case the test setup fails at the dialer:
```
$ make test GINKGO_ARGS="-v --procs=1 --focus='A PlacementAPI instance is created without container images'"
[...snip...]
Will run 1 of 9 specs
------------------------------
[BeforeSuite] 
/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:84
  STEP: bootstrapping test environment @ 07/14/23 10:57:36.494
  2023-07-14T10:57:36+02:00	DEBUG	controller-runtime.test-env	starting control plane
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading Webhooks from path	{"path": "../../config/webhook"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "kustomization.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "kustomizeconfig.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "manifests.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read webhooks from file	{"file": "service.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRDs
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "../../config/crd/bases"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "placement.openstack.org_placementapis.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/keystone-operator/api@v0.0.0-20230622141005-e9220a4b3dfe/bases"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "keystone.openstack.org_keystoneapis.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "keystone.openstack.org_keystoneendpoints.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "keystone.openstack.org_keystoneservices.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/mariadb-operator/api@v0.0.0-20230622153114-756aead1d819/bases"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "mariadb.openstack.org_galeras.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "mariadb.openstack.org_mariadbdatabases.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "mariadb.openstack.org_mariadbs.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	reading CRDs from path	{"path": "/home/gibi/go/pkg/mod/github.com/openstack-k8s-operators/lib-common/modules/test@v0.0.0-20230711124224-54bd95cadcf3/openshift_crds/route/v1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	read CRDs from file	{"file": "route_crd.yaml"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "galeras.mariadb.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "mariadbdatabases.mariadb.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "mariadbs.mariadb.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "routes.route.openshift.io"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "placementapis.placement.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "keystoneapis.keystone.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "keystoneendpoints.keystone.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing CRD	{"crd": "keystoneservices.keystone.openstack.org"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "mariadb.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "mariadb.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "mariadb.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "route.openshift.io/v1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "placement.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "keystone.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "keystone.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	adding API in waitlist	{"GV": "keystone.openstack.org/v1beta1"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing webhooks
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing mutating webhook	{"webhook": "mutating-webhook-configuration"}
  2023-07-14T10:57:39+02:00	DEBUG	controller-runtime.test-env	installing validating webhook	{"webhook": "validating-webhook-configuration"}
  2023-07-14T10:57:39+02:00	INFO	placementapi-resource	PlacementAPI defaults initialized	{"defaults": {"ContainerImageURL":"quay.io/podified-antelope-centos9/openstack-placement-api:current-podified"}}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.PlacementAPI"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.MariaDBDatabase"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.KeystoneService"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1beta1.KeystoneEndpoint"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Job"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Service"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Secret"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.ConfigMap"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Deployment"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Route"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.ServiceAccount"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.Role"}
  2023-07-14T10:57:39+02:00	INFO	Starting EventSource	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "source": "kind source: *v1.RoleBinding"}
  2023-07-14T10:57:39+02:00	INFO	Starting Controller	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI"}
  2023-07-14T10:57:39+02:00	INFO	Starting workers	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI", "worker count": 1}
  [FAILED] in [BeforeSuite] - /home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:197 @ 07/14/23 10:57:40.646
[BeforeSuite] [FAILED] [4.153 seconds]
[BeforeSuite] 
/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:84

  [FAILED] Timed out after 1.001s.
  Expected success, but got an error:
      <*net.OpError | 0xc000559ef0>: 
      dial tcp 127.0.0.1:40081: connect: connection refused
      {
          Op: "dial",
          Net: "tcp",
          Source: nil,
          Addr: <*net.TCPAddr | 0xc000930960>{
              IP: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 127, 0, 0, 1],
              Port: 40081,
              Zone: "",
          },
          Err: <*os.SyscallError | 0xc0001e50c0>{
              Syscall: "connect",
              Err: <syscall.Errno>0x6f,
          },
      }
  In [BeforeSuite] at: /home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:197 @ 07/14/23 10:57:40.646

  Full Stack Trace
    github.com/openstack-k8s-operators/placement-operator/tests/functional_test.glob..func3()
    	/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:197 +0x13c7
------------------------------
[AfterSuite] 
/home/gibi/upstream/git/openstack-k8s-operators/placement-operator/tests/functional/suite_test.go:200
  STEP: tearing down the test environment @ 07/14/23 10:57:40.647
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for non leader election runnables
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for leader election runnables
  2023-07-14T10:57:40+02:00	INFO	Shutdown signal received, waiting for all workers to finish	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI"}
  2023-07-14T10:57:40+02:00	INFO	All workers finished	{"controller": "placementapi", "controllerGroup": "placement.openstack.org", "controllerKind": "PlacementAPI"}
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for caches
  2023-07-14T10:57:40+02:00	INFO	Stopping and waiting for webhooks
  2023-07-14T10:57:40+02:00	INFO	Wait completed, proceeding to shutdown the manager
[AfterSuite] PASSED [1.124 seconds]
------------------------------
[...snip...]
```
so the current check catches a missing webhook.

2. Then if I apply the same diff to the the current PR then the new check passes and only the actual test case fails later when try to call the webhook. 

This shows me that the current check is doing more than the proposed one.

kajinamit replied:
I think this behavior is caused by `GetWebhookServer` which directs the request to controller manager port in case webhook is not registered.
 https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.14/pkg/manager/internal.go#L309-L316

This behavior was changed in 0.15 but we are still using old 0.14 for now.
 https://github.com/kubernetes-sigs/controller-runtime/commit/0377e0f7a41bd9d88d43811192309b193c9cd793

We can probably add some checks to ensure we are sending the requests to the expected server/port, but I'm wondering if it's worth effort here, because
* we are catching the failure even in an ugly way
* this will be improved in the new controller-runtime version

@gibizer May I know your thoughts on this ?

gibizer replied:
Can we wait for the controller-runtime@0.15 bump before we change this? Or the change in main.go fixes an immediate problem with our operators?

kajinamit replied:
I can remove this change now. I thought using the built-in function can simplify the implementation(especially because we are using the same implementation in the actual operator code) but that's not as not urgent as the original problem.

gibizer commented on main.go:
```diff
@@ -115,21 +115,23 @@ func main() {
+	if err := mgr.AddReadyzCheck("readyz", checker); err != nil {
 		setupLog.Error(err, "unable to set up ready check")
 		os.Exit(1)
 	}
```
Is this change fixes an actual bug in the system? If so then I can be convinced to land this before the dev preview.

kajinamit replied:
This fixes the api failures caused by non-ready webhook, which we initially found in edpm jobs.

example:

https://logserver.rdoproject.org/15/215/520a66159a28fba5a4d21271d217e493572436c2/github-check/edpm-ansible-crc-podified-edpm-baremetal/28c43f7/ci-framework-data/logs/ci_make_005_run_edpm_baremetal.log
```
Internal error occurred: failed calling webhook "mnetconfig.kb.io": failed to call webhook: Post "https://infra-operator-controller-manager-service.openstack-operators.svc:443/mutate-network-openstack-org-v1beta1-netconfig?timeout=10s": dial tcp 10.217.0.93:9443: connect: connection refused
```

The above example error is triggered by infra-operator but this problem can occur in all operators now.

gibizer replied:
Thanks for the pointers.
So from the log:
```
++ oc get -n openstack-operators csv/infra-operator.v0.0.1 -o 'jsonpath={.status.phase}'
+ PHASE=Succeeded
+ echo Succeeded
+ '[' Succeeded '!=' Succeeded ']'
+ exit 0
+ rc=0
+ popd
+ exit 0
+ '[' -z /home/zuul/ci-framework-data/artifacts/manifests/openstack/infra/cr ']'
+ NEXT_WAIT_TIME=0
+ '[' 0 -eq 15 ']'
+ oc kustomize /home/zuul/ci-framework-data/artifacts/manifests/openstack/infra/cr
+ oc apply -f -
Error from server (InternalError): error when creating "STDIN": Internal error occurred: failed calling webhook "mnetconfig.kb.io": failed to call webhook: Post "https://infra-operator-controller-manager-service.openstack-operators.svc:443/mutate-network-openstack-org-v1beta1-netconfig?timeout=10s": dial tcp 10.217.0.93:9443: connect: connection refused
```
The script waits for the .status.phase to become "Succeeded" before it tries to use the operator. Will the proposed change prevents the operator to reach "Succeeded" state if the webhooks are not started yet? What will be the phase while the operator waits for the webhooks?

kajinamit replied:
That phase become "Succeeded" after the controller-manager deployment becomes ready and the controller-manage deployment becomes ready when the controller-manager pod is launched and it responds to the healthcheck using the `readyz` healthcheck.

https://github.com/openstack-k8s-operators/placement-operator/blob/main/config/manager/manager.yaml#L40-L51

Current problem is that these healthchecks (both readyz for readiness and healthz for liveness) are doing null check so the pod/deployment becomes ready once it launches the healthcheck endpoints and there is no mechanism to check availability/readiness of webhook.

With this change, readyz healthcheck returns success only after webhook severer becomes ready, and deployment readiness and csv phase is updated only after webhook server is started and becomes ready.

I've also updated the `healthz` check as well because API would not function properly if webhook is down.

gibizer replied:
Thanks for the explanation. I think I finally get the full picture. If we are merging such change across our operators before the dev preview then I will not block this either before the dev preview.

kajinamit replied:
Thanks. I'll bring this to our sync call tomorrow to get attention from the team and hopefully agreement with merging this early as a bug fix.
